recirq/otoc/parallel_xeb.py

```diff
@@ -0,0 +1,725 @@
+        num_points: The total number of XEB fidelities to be used in the
+            cost function for optimization. Default is 8, such that the cost
+            function is the sum of the XEB fidelities for the first 8 numbers
+            of cycles in num_cycle_range.
+        plot_individual_traces: Whether to plot the XEB fidelities along with
+            the fitting results for each qubit pair.
+        plot_histograms: Whether to plot the histograms of cycle Pauli errors
+            and changes in FSIM angles after fitting for all qubit pairs.
+        save_directory: Directory to which the plots are to be saved. If
+            unspecified, the plots will not be saved.
+
+    Returns:
+        fitted_gates: A dictionary with qubit pairs as keys and optimized
+            FSIM unitaries, represented by cirq.Circuit objects, as values.
+        correction_gates: Same as fitted_gates, but with all Z rotations
+            reversed in signs.
+        fitted_angles: A dictionary with qubit pairs as keys and optimized
+            FSIM unitaries as values. Here the FSIM unitaries are represented
+            as a dictionaries with the names of the FSIM phases as keys and
+            their fitted values as values.
+        final_errors_optimized: A dictionary with qubit pairs as keys and
+            their cycle errors after fitting as values.
+        final_errors_unoptimized: A dictionary with qubit pairs as keys and
+            their cycle errors before fitting as values.
+    """
+    num_trials = len(measured_bits[0])
+    p_data_all, sq_gates = _pairwise_xeb_probabilities(
+        all_qubits, num_cycle_range, measured_bits, scrambling_gates,
+        interaction_sequence)
+    final_errors_unoptimized = {}
+    final_errors_optimized = {}
+    delta_angles = {}
+    purity_errors = {}
+    fitted_gates = {}
+    fitted_angles = {}
+    correction_gates = {}
+
+    for (q0, q1), p_data in p_data_all.items():
+        print('Fitting qubits {} and {}'.format(q0, q1))
+
+        def xeb_fidelity(angle_shifts: np.ndarray, num_p: int
+                         ) -> Tuple[float, List[float], np.ndarray,
+                                    np.ndarray, float]:
+            new_angles = fsim_angles.copy()
+            for i, angle_name in enumerate(_fsim_angle_labels):
+                new_angles[angle_name] += angle_shifts[i]
+            fsim_mat = angles_to_fsim(**new_angles)
+
+            max_cycles = num_cycle_range[num_p - 1]
+
+            p_sim = [np.zeros((num_trials, 4)) for _ in
+                     range(max_cycles)]
+            for i in range(num_trials):
+                unitary = np.identity(4, dtype=complex)
+                for j in range(max_cycles):
+                    mat_0 = _rot_mats[sq_gates[(q0, q1)][i][0, j]]
+                    mat_1 = _rot_mats[sq_gates[(q0, q1)][i][1, j]]
+                    unitary = np.kron(mat_0, mat_1).dot(unitary)
```

Comment:
is this implementing a full statevector simulator within this closure?? Consider using `cirq.Circuit` and the built in cirq simulator functionality.

Reply:
Yeah, I tried that in the beginning. It was significantly slower than simply multiplying matrices here. Cirq simulator appears to only be faster when the number of qubits go up to 8 or so. Maybe the simulator has changed since then (~1 year ago), but I don't see much harm keeping this the way it is. It's just two qubits after all.